### PR TITLE
[ONNX] Added hardswish symbolic in opset 9

### LINKED ISSUE #43665: Make torch.nn.Hardswish ONNX exportable
## 🐛 Bug

torch.nn.Hardswish is not ONNX exportable

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

### PR DESCRIPTION
**Pull request by shubhambhokare1**
Adds support for torch.nn.Hardswish operator in Export

Fixes #43665

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -1742,6 +1742,15 @@ def hardtanh(g, self, min_val, max_val):
     return g.op("Clip", self, min_f=min_val, max_f=max_val)
 
 
+@parse_args('v')
+def hardswish(g, self):
```
Comment (spandantiwari) [inline torch/onnx/symbolic_opset9.py:1746]: Is it added for older opsets 7 and 8 (assuming they support all the needed ops)? If not then we should ensure there's a proper error message coming out for that (I think we already have a list for ops the are not supported in 7 and 8).

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -221,6 +221,13 @@ def _slice_helper(g, input, axes, starts, ends, steps=None, dynamic_slice=False)
         from torch.onnx.symbolic_opset10 import _slice
         return _slice(g, input, axes, starts, ends, steps, dynamic_slice)
 
+def _hardtanh_helper(g, input, min_val, max_val):
+    if _export_onnx_opset_version <= 10:
+        from torch.onnx.symbolic_opset9 import hardtanh
+        return hardtanh(g, input, min_val, max_val)
+    else:
+        from torch.onnx.symbolic_opset11 import hardtanh
```
[the remark this answers is not in the document]
Comment (shubhambhokare1) [reply torch/onnx/symbolic_helper.py:229]: (1) There exists a hardtanh symbolic in opset 11
(2) The point of this branch is utilize the correct hardtanh symbolic based on opset versions

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (shubhambhokare1) [reply torch/onnx/symbolic_opset9.py:1746]: Yes, the test cases for opset 7 and 8 are passing

File: torch/onnx/symbolic_helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/onnx/symbolic_helper.py:229]: nit: title -> Added hardswish symbolic for opset 9?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -586,6 +586,27 @@ def forward(self, x):
         x = torch.arange(-5, 5).to(dtype=torch.float32)
         self.run_test(MyModel(), x)
 
+    def test_hardswish(self):
```
Comment (spandantiwari) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:589]: Can we add this op in the list of supported ops on the doc page: https://pytorch.org/docs/master/onnx.html?

**Review by spandantiwari — APPROVED** [review]
LGTM.

**Review by bzinodev — APPROVED** [review]

**Comment by Adhithya-tech** [comment]
Could you please tell me how to get it to work when mobilenetv3 is used as backbone in FasterRCNN ?

**Comment by zhiqwang** [comment]
Hi @Adhithya-tech, torch 1.8.0 should supported exporting to onnx with the Hardswich ops, the error in mobilenetv3 is arose from Hardsigmoid, and #54193 is working on this. [Here](https://github.com/pytorch/vision/issues/3463#issuecomment-791191499) is a temporary solution, maybe you could check it for more info.

**Comment by Adhithya-tech** [comment]
@zhiqwang thank you for the reply, but if you don't mind could you point towards any tutorials to replace the hardsigmoid for a pre-trained model?

**Comment by zhiqwang** [comment]
Hi @Adhithya-tech , you can just replace `return F.hardsigmoid(scale, inplace=inplace)` to `return F.hardtanh(scale + 3, 0., 6., inplace=inplace) / 6.` in [here](https://github.com/pytorch/vision/blob/3bfdb4275c13ec8207ae78222b7895c01a01371b/torchvision/models/mobilenetv3.py#L35) for a temporary solution, this two ops are numerically equal.

**Comment by Adhithya-tech** [comment]
Thank you for the swift reply

**Comment by Adhithya-tech** [comment]
How about for a faster RCNN model that already has a pretrained mobilenet backbone?

**Comment by zhiqwang** [comment]
The operation of transferring to ONNX should be the same as that of resnet50 backbone.

**Comment by Adhithya-tech** [comment]
My brain seems too dead to understand what you're saying. How can can I attach the changed mobilenet backbone to a pretrained faster RCNN model?
If what you mean by the above comment is, why not use a resnet-50 backbone----Then I would say that I need smaller model and much faster inference time.

**Comment by zhiqwang** [comment]
Oops, what I want to say is that you can do it the same as when you exporting a ResNet50-based faster RCNN model to ONNX. Just replace the Hardsigmoid ops, or you can wait for #54193 to be merged.

**Comment by Adhithya-tech** [comment]
Got it, thanks

**Comment by PeterL1n** [comment]
@zhiqwang There is a ONNX op for HardSigmoid since opset 6. And there will a dedicated HardSwish op since opset 14.

For now, why not export it as x * HardSigmoid(x) instead. It is cleaner in the graph. Currently it has a lot of separate add, clip, div, operations, which can be simplified as a single HardSigmoid operation.

#54193 has already been merged, and it is implemented using the HardSigmoid op. Please take a look.

**Comment by zhiqwang** [comment]
> For now, why not export it as x * HardSigmoid(x) instead. It is cleaner in the graph. Currently it has a lot of separate add, clip, div, operations, which can be simplified as a single HardSigmoid operation.

Hi @PeterL1n , I second your proposal here :) Maybe it would be better to raise a new issue about it?

**Comment by PeterL1n** [comment]
@zhiqwang Sounds good. I am not familiar with pytorch development, but it should be a pretty simple change. If you can implement it that would be awesome, or if you only plan to create a new issue for someone else to look at, that's great too.

**Pull request closed without merging**